File: main-service/src/main/java/ru/pgk/main_service/features/teacher/repositoties/TeacherRepository.java
package ru.pgk.main_service.features.teacher.repositoties;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import ru.pgk.main_service.features.teacher.entities.TeacherEntity;

import java.util.Optional;

public interface TeacherRepository extends JpaRepository<TeacherEntity, Integer> {

    @Query(
            "SELECT teacher FROM teachers teacher " +
                    "WHERE LOWER(teacher.lastName) LIKE %?1% " +
                    "OR LOWER(teacher.firstName) LIKE %?1% OR " +
                    "LOWER(teacher.middleName) LIKE %?1%"
    )
    Page<TeacherEntity> search(String name, Pageable pageable);

    Optional<TeacherEntity> findByCabinet(String cabinet);
}

File: teacher-service/src/main/kotlin/ru/pgk/teacher_service/features/repositories/TeacherRepository.kt
package ru.pgk.teacher_service.features.repositories

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import ru.pgk.teacher_service.features.entities.TeacherEntity
import java.util.*

interface TeacherRepository: JpaRepository<TeacherEntity, Int> {

    @Query(
        "SELECT teacher FROM teachers teacher " +
                "WHERE LOWER(teacher.lastName) LIKE %?1% " +
                "OR LOWER(teacher.firstName) LIKE %?1% OR " +
                "LOWER(teacher.middleName) LIKE %?1%"
    )
    fun search(name: String, pageable: Pageable): Page<TeacherEntity>

    fun findByCabinet(cabinet: String): Optional<TeacherEntity>
}